Neos.Flow/Tests/Unit/Http/Middleware/MethodOverrideMiddlewareTest.php

```diff
@@ -0,0 +1,107 @@
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Http\Middleware\MethodOverrideMiddleware;
+use Neos\Flow\Tests\UnitTestCase;
+use PHPUnit\Framework\MockObject\MockObject;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class MethodOverrideMiddlewareTest extends UnitTestCase
+{
+
+    private MethodOverrideMiddleware $middleware;
+
+    /**
+     * @var ServerRequestInterface|MockObject
+     */
+    private $mockRequest;
+
+    /**
+     * @var RequestHandlerInterface|MockObject
+     */
+    private $mockRequestHandler;
+
+    /**
+     * @var ResponseInterface|MockObject
+     */
+    private $mockResponse;
+
+    public function setUp(): void
+    {
+        $this->middleware = new MethodOverrideMiddleware();
+
+        $this->mockRequest = $this->getMockBuilder(ServerRequestInterface::class)->getMock();
+        $this->mockRequestHandler = $this->getMockBuilder(RequestHandlerInterface::class)->getMock();
+        $this->mockResponse = $this->getMockBuilder(ResponseInterface::class)->getMock();
+    }
+
+    public function process_matchingRequests_dataProvider(): \Traversable
+    {
+        yield 'parsedBody (__method)' => ['method' => 'POST', 'headers' => [], 'parsedBody' => ['__method' => 'PUT'], 'expectedMethod' => 'PUT'];
+        yield 'header (X-Http-Method-Override)' => ['method' => 'POST', 'headers' => ['X-Http-Method-Override' => 'PATCH'], 'parsedBody' => [], 'expectedMethod' => 'PATCH'];
+        yield 'header (X-Http-Method header)' => ['method' => 'POST', 'headers' => ['X-Http-Method' => 'DELETE'], 'parsedBody' => [], 'expectedMethod' => 'DELETE'];
+        yield 'parsedBody and X-Http-Method header' => ['method' => 'POST', 'headers' => ['X-Http-Method' => 'DELETE'], 'parsedBody' => ['__method' => 'PUT'], 'expectedMethod' => 'PUT'];
+        yield 'X-Http-Method-Override and X-Http-Method header' => ['method' => 'POST', 'headers' => ['X-Http-Method-Override' => 'PATCH', 'X-Http-Method' => 'DELETE'], 'parsedBody' => [], 'expectedMethod' => 'PATCH'];
+        yield 'parsedBody and both headers' => ['method' => 'POST', 'headers' => ['X-Http-Method-Override' => 'PATCH', 'X-Http-Method' => 'DELETE'], 'parsedBody' => ['__method' => 'PUT'], 'expectedMethod' => 'PUT'];
+    }
+
+    /**
+     * @test
+     * @dataProvider process_matchingRequests_dataProvider
+     */
+    public function process_matchingRequests(string $method, array $headers, $parsedBody, string $expectedMethod): void
+    {
+        $this->mockRequest->method('getMethod')->willReturn($method);
+        $this->mockRequest->method('getParsedBody')->willReturn($parsedBody);
+        $this->mockRequest->method('hasHeader')->willReturnCallback(fn($header) => isset($headers[$header]));
+        $this->mockRequest->method('getHeaderLine')->willReturnCallback(fn($header) => $headers[$header]);
```

Comment:
When we're at it we could as well extract this part into a `prepareMockRequest(string $method, array $headers, $parsedBody)` method to make the tests themself more readable. I always found our tests way too hard to decypher because we never cared to replace PHPUnit technicalities with meaningful patterns.